src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -6,13 +6,25 @@ function frule((_, Δx, Δy), ::typeof(dot), x, y)
     return dot(x, y), dot(Δx, y) + dot(x, Δy)
 end
 
-function rrule(::typeof(dot), x, y)
+function rrule(::typeof(dot), x::AbstractArray, y::AbstractArray)
     function dot_pullback(ΔΩ)
-        return (NoTangent(), @thunk(y .* ΔΩ'), @thunk(x .* ΔΩ))
+        xthunk = InplaceableThunk(
+            @thunk(reshape(y .* ΔΩ', axes(x))),
+            dx -> dx .+= reshape(y, axes(x)) .* ΔΩ',
```

Comment:
This feels like something that should have a fast BLAS overload, but having checked, it doesn't.
Julia code should be pretty good with fusing anyway.

(not blocking)

Reply:
It does reshape sometimes when not strictly necessary. But I didn't see a good way to avoid that.

Reply:
reshape is nonallocating anyway, it is just a view.

Reply:
Sometimes O(1) allocations, but never O(length). But I meant more things like this:
```
julia> gradient(x->dot(Diagonal([1,2,3]),x), rand(3,3))[1]
3×3 reshape(::Diagonal{Float64, Vector{Float64}}, 3, 3) with eltype Float64:
 1.0  0.0  0.0
 0.0  2.0  0.0
 0.0  0.0  3.0
```
vs before
```
julia> gradient(x->dot(Diagonal([1,2,3]),x), rand(3,3))[1]
3×3 Diagonal{Float64, Vector{Float64}}:
 1.0   ⋅    ⋅ 
  ⋅   2.0   ⋅ 
  ⋅    ⋅   3.0
```